### PR TITLE
merge task2 to main

### PR DESCRIPTION
**Pull request by igor-zhuravlev**

### CODE REVIEW

File: task2/rest-api-basics/.gitattributes [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#
```
Comment (AliaksandrLiulis) [inline task2/rest-api-basics/.gitattributes]: @igor-zhuravlev extra file

File: task2/rest-api-basics/build.gradle [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        withJavadocJar()
+    }
+
+    repositories {
+        jcenter()
+        mavenCentral()
+    }
+
+    dependencies {
+        implementation group: 'org.apache.logging.log4j', name: 'log4j-core', version: '2.14.0'
```
Comment (AliaksandrLiulis) [inline task2/rest-api-basics/build.gradle]: @igor-zhuravlev Could you move all version numbers to const.

File: task2/rest-api-basics/domain/src/main/java/com/epam/esm/entity/GiftCertificate.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+import java.io.Serializable;
+import java.math.BigDecimal;
+import java.time.LocalDateTime;
+import java.util.Objects;
+import java.util.Set;
+
+public class GiftCertificate implements Serializable {
+    private static final long serialVersionUID = 7468316931994434280L;
+
+    private Long id;
```
Comment (AliaksandrLiulis) [inline task2/rest-api-basics/domain/src/main/java/com/epam/esm/entity/GiftCertificate.java]: @igor-zhuravlev The same (Certificate, Tag) id and name fields. Could be moved to abstract class.

File: task2/rest-api-basics/gradlew [hunk 4]
```diff
@@ -0,0 +1,185 @@
+#!/usr/bin/env sh
```
Comment (AliaksandrLiulis) [inline task2/rest-api-basics/gradlew]: @igor-zhuravlev extra file

File: task2/rest-api-basics/gradlew.bat [hunk 5]
```diff
@@ -0,0 +1,89 @@
+@rem
```
Comment (AliaksandrLiulis) [inline task2/rest-api-basics/gradlew.bat]: @igor-zhuravlev move all extra files to git ignore

Comment (AliaksandrLiulis) [reply task2/rest-api-basics/gradlew.bat]: @igor-zhuravlev also you have some empty files like task2/rest-api-basics/profile-dev.gradle

File: task2/rest-api-basics/repository/build.gradle [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+            password dbProperties.get('jdbc.password')
+        }
+    }
+    runList = 'main'
+}
+
+dependencies {
+    implementation project(':domain')
+
+    implementation group: 'org.postgresql', name: 'postgresql', version: '42.2.18'
```
Comment (AliaksandrLiulis) [inline task2/rest-api-basics/repository/build.gradle]: @igor-zhuravlev Could you move all version numbers to const

File: task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/GiftCertificateRepository.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+import org.springframework.data.domain.Sort;
+
+import java.util.List;
+import java.util.Set;
+
+public interface GiftCertificateRepository extends Repository<GiftCertificate, Long> {
+    GiftCertificate findByName(String name) throws RepositoryException;
+    List<GiftCertificate> findByTagName(String tagName) throws RepositoryException;
+    Long updateById(Long id, GiftCertificate giftCertificate) throws RepositoryException;
+    Long deleteByName(String name) throws RepositoryException;
```
Comment (AliaksandrLiulis) [inline task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/GiftCertificateRepository.java]: @igor-zhuravlev can the tags contain the same names? Why not by id?

File: task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/GiftCertificateRepository.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+    Set<GiftCertificate> findAllWithTags() throws RepositoryException;
```
Comment (AliaksandrLiulis) [inline task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/GiftCertificateRepository.java]: @igor-zhuravlev Why do you need so many identical methods? Can you define the execution logic at the service level, generate a request, and call one method? There are a lot of unnecessary methods.

File: task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+public class GiftCertificateRepositoryImpl implements GiftCertificateRepository {
+    private static final Logger logger = LogManager.getLogger(GiftCertificateRepositoryImpl.class);
+
+    @Autowired
+    private JdbcTemplate jdbcTemplate;
+
+    private static final String FIND_ALL_QUERY =
+            "SELECT id gc_id, name gc_name, description, price, duration, create_date, last_update_date " +
+            "FROM gift_certificate";
+    private static final String FIND_ALL_WITH_TAGS_QUERY =
```
Comment (AliaksandrLiulis) [inline task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java]: @igor-zhuravlev A lot of copypast.  Your queries are almost the same in terms of parameters and differ only in search and sorting parameters. Why not generate all this in one place?

File: task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+                    UPDATE_BY_ID_QUERY,
+                    giftCertificate.getName(),
+                    giftCertificate.getDescription(),
+                    giftCertificate.getPrice(),
+                    giftCertificate.getDuration(),
+                    giftCertificate.getLastUpdateDate(),
+                    id
+            );
+
+            if (giftCertificate.getTags() != null) {
```
Comment (AliaksandrLiulis) [inline task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java]: @igor-zhuravlev All logic to service layer.

File: task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+            "duration = COALESCE(?, duration), " +
+            "last_update_date = ? " +
+            "WHERE id = ?";
+
+    @Override
+    public GiftCertificate findById(Long id) throws RepositoryException {
+        try {
+            return jdbcTemplate.queryForObject(FIND_BY_ID_QUERY, new GiftCertificateMapper(), id);
+        } catch (EmptyResultDataAccessException e) {
+            return null;
```
Comment (AliaksandrLiulis) [inline task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java]: @igor-zhuravlev What does it mean? You can apply ifexist method end return exception with message

File: task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java [hunk 12]
```diff
@@ -0,0 +1,321 @@
...
+            throw new RepositoryException(e);
+        }
+    }
+
+    @Override
+    public GiftCertificate findByName(String name) throws RepositoryException {
+        try {
+            return jdbcTemplate.queryForObject(FIND_BY_NAME_QUERY, new GiftCertificateMapper(), name);
+        } catch (EmptyResultDataAccessException e) {
+            return null;
```
Comment (AliaksandrLiulis) [inline task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java]: @igor-zhuravlev The same

File: task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/TagRepositoryImpl.java [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+            tag.setId((Long) keyHolder.getKeys().get(TagMapper.SECONDARY_ID));
+
+            return tag;
+        } catch (DataAccessException e) {
+            throw new RepositoryException(e);
+        }
+    }
+
+    @Override
+    public Long deleteByName(String name) throws RepositoryException {
```
Comment (AliaksandrLiulis) [inline task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/TagRepositoryImpl.java]: @igor-zhuravlev Why not tag id. It's unique.

Comment (AliaksandrLiulis) [reply task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/TagRepositoryImpl.java]: @igor-zhuravlev if you pass a name from above, just check against the existing name. If exists, return, but not, throw exception

File: task2/rest-api-basics/repository/src/main/resources/liquibase/db.changelog-master.xml [hunk 14]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (AliaksandrLiulis) [inline task2/rest-api-basics/repository/src/main/resources/liquibase/db.changelog-master.xml]: @extra file

File: task2/rest-api-basics/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+    @Autowired
+    private GiftCertificateRepository giftCertificateRepository;
+    @Autowired
+    private TagRepository tagRepository;
+    @Autowired
+    private Converter<GiftCertificate, GiftCertificateDto> giftCertificateConverter;
+
+    @Override
+    public List<GiftCertificateDto> findAllWithTags(Map<String, String[]> params) throws ServiceException {
+        try {
```
Comment (AliaksandrLiulis) [inline task2/rest-api-basics/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: @igor-zhuravlev It would be nice to use this construct elsewhere which defines the calling method

File: task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+import java.util.stream.Collectors;
+
+@RestController
+public class GiftCertificateController {
+    private static final Logger logger = LogManager.getLogger(GiftCertificateController.class);
+
+    @Autowired
+    private GiftCertificateService giftCertificateService;
+
+    private Map<String, String[]> multiValueMapToMap(MultiValueMap<String, String> requestParams) {
```
Comment (AliaksandrLiulis) [inline task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java:32]: @igor-zhuravlev Strange construction.

File: task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+        return requestParams.entrySet().stream()
+                .collect(Collectors.toMap(
+                        Map.Entry::getKey,
+                        e -> e.getValue().toArray(String[]::new)
+                ));
+    }
+
+    @GetMapping("/gifts-tags")
```
Comment (AliaksandrLiulis) [inline task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java]: @igor-zhuravlev Would be better underscore

File: task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+    public GiftCertificateDto saveGiftCertificate(@RequestBody GiftCertificateDto giftCertificateDto) {
+        try {
+            return giftCertificateService.save(giftCertificateDto);
+        } catch (ServiceException e) {
+            logger.error(e);
+            throw e;
+        }
+    }
+
+    @PutMapping("/gifts/{name}")
```
Comment (AliaksandrLiulis) [inline task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java]: @igor-zhuravlev Would be better by id. Certifictes can be with same names.

File: task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+    public GiftCertificateDto updateGiftCertificate(@PathVariable String name, @RequestBody GiftCertificateDto giftCertificateDto) {
+        try {
+            return giftCertificateService.updateByName(name, giftCertificateDto);
+        } catch (ServiceException e) {
+            logger.error(e);
+            throw e;
+        }
+    }
+
+    @DeleteMapping("/gifts/{name}")
```
Comment (AliaksandrLiulis) [inline task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java]: @igor-zhuravlev id

File: task2/rest-api-basics/domain/src/main/java/com/epam/esm/entity/GiftCertificate.java [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+    private String name;
+    private String description;
+    private BigDecimal price;
+    private Integer duration;
+    private LocalDateTime createDate;
```
Comment (mvladmoss) [inline task2/rest-api-basics/domain/src/main/java/com/epam/esm/entity/GiftCertificate.java]: Suggest to use Instant instead of LocalDateTime

File: task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/mapper/GiftCertificateMapper.java [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+    public static final String ID = "gc_id";
+    public static final String NAME = "gc_name";
+    public static final String DESCRIPTION = "description";
+    public static final String PRICE = "price";
+    public static final String DURATION = "duration";
+    public static final String CREATE_DATE = "create_date";
+    public static final String LAST_UPDATE_DATE = "last_update_date";
+
+    public static final String SECONDARY_ID = "id";
+    public static final String SECONDARY_NAME = "name";
```
Comment (mvladmoss) [inline task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/mapper/GiftCertificateMapper.java]: unused field

File: task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/mapper/TagMapper.java [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+public class TagMapper implements RowMapper<Tag> {
+
+    public static final String ID = "t_id";
+    public static final String NAME = "t_name";
+
+    public static final String SECONDARY_ID = "id";
+    public static final String SECONDARY_NAME = "name";
```
Comment (mvladmoss) [inline task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/mapper/TagMapper.java]: unused field

File: task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java [hunk 23]
```diff
@@ -0,0 +1,321 @@
...
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Repository;
+
+import java.sql.PreparedStatement;
+import java.sql.Timestamp;
+import java.util.*;
+
+@Repository
+public class GiftCertificateRepositoryImpl implements GiftCertificateRepository {
+    private static final Logger logger = LogManager.getLogger(GiftCertificateRepositoryImpl.class);
```
Comment (mvladmoss) [inline task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java]: The logger is declared but is not used anywhere

File: task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java (hunk 11 above)
Comment (mvladmoss) [reply task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java]: I can suggest using the Optional class for such purposes.

File: task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java [hunk 24]
```diff
@@ -0,0 +1,321 @@
...
+                return preparedStatement;
+            };
+            return jdbcTemplate.query(preparedStatementCreator, new GiftCertificateMapper());
+        } catch (DataAccessException e) {
+            throw new RepositoryException(e);
+        }
+    }
+
+    @Override
+    public Long updateById(Long id, GiftCertificate giftCertificate) throws RepositoryException {
```
Comment (mvladmoss) [inline task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java]: Firstly, I suggest renaming it to just "update". Secondly, why do you pass both id and certificate? As I see your GiftCertificate class includes the id field.

File: task2/rest-api-basics/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,230 @@
...
+import org.springframework.data.domain.Sort;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.time.LocalDateTime;
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Service
+public class GiftCertificateServiceImpl implements GiftCertificateService {
```
Comment (mvladmoss) [inline task2/rest-api-basics/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:37]: Please use @Transactiona(readOnly = true) annotation for read operations too.
image
https://stackoverflow.com/questions/44984781/what-are-advantages-of-using-transactionalreadonly-true

File: task2/rest-api-basics/web/src/main/java/com/epam/esm/web/config/WebAppConfig.java [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+    @Bean
+    public HikariDataSource hikariDataSource() {
+        HikariConfig hikariConfig = new HikariConfig();
+        hikariConfig.setDataSource(dataSource());
+        HikariDataSource hikariDataSource = new HikariDataSource(hikariConfig);
+        return hikariDataSource;
+    }
+
+    @Bean
+    public DataSource dataSource() {
```
Comment (mvladmoss) [inline task2/rest-api-basics/web/src/main/java/com/epam/esm/web/config/WebAppConfig.java]: What does configuration which is related to the database do in WebConfig?

File: task2/rest-api-basics/web/src/main/java/com/epam/esm/web/config/WebAppConfig.java [hunk 27]
```diff
@@ -0,0 +1,73 @@
...
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.datasource.DriverManagerDataSource;
+import org.springframework.jdbc.support.JdbcTransactionManager;
+import org.springframework.transaction.PlatformTransactionManager;
+import org.springframework.transaction.annotation.EnableTransactionManagement;
+
+import javax.sql.DataSource;
+
+@Configuration
+@EnableTransactionManagement
```
Comment (mvladmoss) [inline task2/rest-api-basics/web/src/main/java/com/epam/esm/web/config/WebAppConfig.java]: Each of your modules(except domain) should include its own **Configuration** class. Each of these configurations has to make component scan only for this module and customize appropriate beans. In order to include configuration from one module to another one, use annotation @Import.

File: task2/rest-api-basics/web/src/main/java/com/epam/esm/web/config/WebAppConfig.java [hunk 28]
```diff
@@ -0,0 +1,73 @@
...
+@PropertySource("classpath:database.properties")
+@PropertySource("classpath:application.properties")
+public class WebAppConfig {
+
+    @Autowired
+    private Environment environment;
+
+    @Bean
+    public PlatformTransactionManager platformTransactionManager() {
+        JdbcTransactionManager jdbcTransactionManager = new JdbcTransactionManager(hikariDataSource());
```
Comment (mvladmoss) [inline task2/rest-api-basics/web/src/main/java/com/epam/esm/web/config/WebAppConfig.java]: inline here and beyond

File: task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java [hunk 29]
```diff
@@ -0,0 +1,101 @@
...
+        try {
+            return giftCertificateService.findAllWithTags(multiValueMapToMap(requestParams));
+        } catch (ServiceException e) {
+            logger.error(e);
+            throw e;
+        }
+    }
+
+
+    @GetMapping("/gifts")
```
Comment (mvladmoss) [inline task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java]: you can put this prefix on the class level because it's duplicated in all endpoints

File: task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java [hunk 30]
```diff
@@ -0,0 +1,101 @@
...
+    private Map<String, String[]> multiValueMapToMap(MultiValueMap<String, String> requestParams) {
+        return requestParams.entrySet().stream()
+                .collect(Collectors.toMap(
+                        Map.Entry::getKey,
+                        e -> e.getValue().toArray(String[]::new)
+                ));
+    }
+
+    @GetMapping("/gifts-tags")
+    public List<GiftCertificateDto> findAllGiftCertificatesWithTags(@RequestParam(required = false) MultiValueMap<String, String> requestParams) {
```
Comment (mvladmoss) [inline task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java]: findAllGiftCertificatesWithTags and findAllGiftCertificates, what is the reason to divide these endpoints?

File: task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java [hunk 31]
```diff
@@ -0,0 +1,101 @@
...
+        try {
+            return giftCertificateService.findAllWithTags(multiValueMapToMap(requestParams));
+        } catch (ServiceException e) {
+            logger.error(e);
```
Comment (mvladmoss) [inline task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java]: For this purpose, you created ExceptionHandler, move logging there

File: task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java [hunk 32]
```diff
@@ -0,0 +1,101 @@
...
+            throw e;
```
Comment (mvladmoss) [inline task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java]: You rethrow ServiceException but it's not handled in ExceptionHandler.

**Comment by mvladmoss** [comment]
Igor, why did you resolve discussions? I don't see any commits that fixed the aforementioned remarks or your comments under these remarks.

**Comment by igor-zhuravlev** [comment]
> Igor, why did you resolve discussions? I don't see any commits that fixed the aforementioned remarks or your comments under these remarks.

Sorry, I collapsed they for myself readability

**Comment by mvladmoss** [comment]
> > Igor, why did you resolve discussions? I don't see any commits that fixed the aforementioned remarks or your comments under these remarks.
> 
> Sorry, I collapsed they for myself readability

Please don't do this

File: task2/rest-api-basics/repository/src/test/java/com/epam/esm/repository/config/RepositoryTestConfig.java [hunk 33]
```diff
@@ -0,0 +1,72 @@
...
+import org.springframework.jdbc.support.JdbcTransactionManager;
+import org.springframework.transaction.PlatformTransactionManager;
+import org.springframework.transaction.annotation.EnableTransactionManagement;
+
+import javax.sql.DataSource;
+
+@Configuration
+@EnableTransactionManagement
+@PropertySource("classpath:database.properties")
+@ComponentScan("com.epam.esm.repository")
```
Comment (mvladmoss) [inline task2/rest-api-basics/repository/src/test/java/com/epam/esm/repository/config/RepositoryTestConfig.java]: I suppose if you try to run your test with this config you will face something like NotUniqueBeanException. In this configuration, you created DataSource bean and simultaneously scan the repository package where another Repository configuration with its own DataSource bean exists.

File: task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java (hunk 16 above)
Comment (igor-zhuravlev) [reply task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java:20]: It convert request query params to Map<K, List<V>>, because i have the same params like sort (&sort=name,asc&sort=date,desc)

File: task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java (hunk 32 above)
Comment (igor-zhuravlev) [reply task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java]: fixed

File: task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java (hunk 31 above)
Comment (igor-zhuravlev) [reply task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java]: fixed

File: task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java (hunk 30 above)
Comment (igor-zhuravlev) [reply task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java]: fixed

File: task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java (hunk 29 above)
Comment (igor-zhuravlev) [reply task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java]: fixed

File: task2/rest-api-basics/web/src/main/java/com/epam/esm/web/config/WebAppConfig.java (hunk 28 above)
Comment (igor-zhuravlev) [reply task2/rest-api-basics/web/src/main/java/com/epam/esm/web/config/WebAppConfig.java]: fixed

File: task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java (hunk 24 above)
Comment (igor-zhuravlev) [reply task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java]: fixed

File: task2/rest-api-basics/repository/src/test/java/com/epam/esm/repository/config/RepositoryTestConfig.java (hunk 33 above)
Comment (igor-zhuravlev) [reply task2/rest-api-basics/repository/src/test/java/com/epam/esm/repository/config/RepositoryTestConfig.java]: Thanks, it fixed.

File: task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/mapper/TagMapper.java [hunk 34]
```diff
@@ -0,0 +1,25 @@
...
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+@Component
+public class TagMapper implements RowMapper<Tag> {
+
+    public static final String ID = "t_id";
+    public static final String NAME = "t_name";
+
+    public static final String SECONDARY_ID = "id";
```
Comment (mvladmoss) [inline task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/mapper/TagMapper.java:16]: Delete unused fields everywhere

**Review by mvladmoss — APPROVED** [review]

File: task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/mapper/TagMapper.java (hunk 34 above)
Comment (igor-zhuravlev) [reply task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/mapper/TagMapper.java:16]: > Delete unused fields everywhere

But it is used in https://github.com/igor-zhuravlev/mjc/blob/c9cb85a679e210b42e235310c95177045a8b78c6/task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/TagRepositoryImpl.java#L74 when I recieve a generated id

Comment (mvladmoss) [reply task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/mapper/TagMapper.java:16]: I didn't notice. Resolved

Comment (igor-zhuravlev) [reply task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/mapper/TagMapper.java:16]: > I didn't notice. Resolved

Sorry for strange question, but how I can resolve other conversations? I need a commented their as fixed or maybe simply click resolve

Comment (mvladmoss) [reply task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/mapper/TagMapper.java:16]: just resolve them

File: task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java (hunk 11 above)
Comment (igor-zhuravlev) [reply task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java]: It throws anyway if not found, thats why I had used catch block

File: task2/rest-api-basics/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java (hunk 25 above)
Comment (igor-zhuravlev) [reply task2/rest-api-basics/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:37]: thanks

File: task2/rest-api-basics/web/src/main/java/com/epam/esm/web/config/WebAppConfig.java (hunk 27 above)
Comment (igor-zhuravlev) [reply task2/rest-api-basics/web/src/main/java/com/epam/esm/web/config/WebAppConfig.java]: thanks

File: task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java (hunk 16 above)
Comment (igor-zhuravlev) [reply task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java:32]: > @igor-zhuravlev Strange construction.

It convert request params to Map<String, List<>>, because I have the same params like sort (&sort=name,asc&sort=date,desc)

**Review by AliaksandrLiulis — APPROVED** [review]

File: task2/rest-api-basics/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 35]
```diff
@@ -0,0 +1,181 @@
...
+import com.epam.esm.service.util.ParamsUtil;
+import com.epam.esm.service.validation.ParamsValidator;
+import com.epam.esm.service.validation.Validator;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.*;
```
Comment (nikskonda) [inline task2/rest-api-basics/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: How many classes from java.util do you use?

File: task2/rest-api-basics/service/src/test/java/com/epam/esm/service/impl/TagServiceImplTest.java [hunk 36]
```diff
@@ -0,0 +1,325 @@
...
+                .should(never())
+                .save(any(Tag.class));
+
+        then(tagConverter)
+                .should(never())
+                .entityToDto(any(Tag.class));
+    }
+
+    @Test
+    void save_ThrowTagNotValidException() {
```
Comment (nikskonda) [inline task2/rest-api-basics/service/src/test/java/com/epam/esm/service/impl/TagServiceImplTest.java]: Just FYI https://dzone.com/articles/7-popular-unit-test-naming

File: task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/TagController.java [hunk 37]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    /**
+     * Saves the tag
+     * @param tagDto - received tag dto
+     * @return found tag dto
+     */
+
+    @PostMapping
+    public TagDto saveTag(@RequestBody TagDto tagDto) {
```
Comment (nikskonda) [inline task2/rest-api-basics/web/src/main/java/com/epam/esm/web/controller/TagController.java]: I think its better to use "save", "find", ... method name, because its obvious that it will be tag from the name of class.

File: task2/rest-api-basics/service/src/main/java/com/epam/esm/service/constant/ServiceError.java [hunk 38]
```diff
@@ -0,0 +1,43 @@
...
+
+    ServiceError(String code) {
+        this.code = code;
+    }
+
+    public String getCode() {
+        return code;
+    }
+
+    private static class Constants {
```
Comment (nikskonda) [inline task2/rest-api-basics/service/src/main/java/com/epam/esm/service/constant/ServiceError.java]: Why did you move error codes to a separate class?

File: task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/config/RepositoryConfig.java [hunk 39]
```diff
@@ -27,7 +27,7 @@ public JdbcTemplate jdbcTemplate() {
         return jdbcTemplate;
     }
 
-    @Bean
+    @Bean(destroyMethod = "close")
```
Comment (nikskonda) [inline task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/config/RepositoryConfig.java]: What is the reason for using destroyMethod = "close"?

Comment (igor-zhuravlev) [reply task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/config/RepositoryConfig.java]: > What is the reason for using destroyMethod = "close"?

I thought if Hikari not autoclosable that put close on bean. Should I delete this?

Comment (nikskonda) [reply task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/config/RepositoryConfig.java]: As I understand, close or shutdown methods will be automatically marked as destroyMethod.
image

Comment (igor-zhuravlev) [reply task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/config/RepositoryConfig.java]: > As I understand, close or shutdown methods will be automatically marked as destroyMethod.
> image

Thank you for your attention. I will fix it.

**Review by nikskonda — APPROVED** [review]
